### PR TITLE
ChibiOS: added warnings for incorrect 1M/2M firmware

### PR DESCRIPTION
**Pull request by tridge**
This adds a warning when running a firmware meant for 2M boards on a board with 1M flash. This can happen if you have an old enough bootloader that it doesn't limit flash size to 1M. This is sent at level CRITICAL so it displays on the HUD:
image
the PR also adds an INFO warning every 30s if you are running a 1M firmware on a board that can use 2M. Sending as INFO means it just shows in the messages tab:
image

### CODE REVIEW

**Comment by IamPete1** [comment]
I'm a little surprised its possible to get a 2M build onto a 1M board and have it work well enough to print a warning.

**Comment by tridge** [comment]
> Have we lost protections on the old VRBrain boards (particularly the v54?) after this?

i've no idea what you're talking about

File: libraries/AP_Vehicle/AP_Vehicle.h [hunk 1]
```diff
@@ -444,6 +444,10 @@ class AP_Vehicle : public AP_HAL::HAL::Callbacks {
     // run notch update at either loop rate or 200Hz
     void update_dynamic_notch_at_specified_rate();
 
+    // decimation for 1Hz update
+    uint8_t one_Hz_counter;
+    void one_Hz_update();
```
[the remark this answers is not in the document]
Comment (tridge) [reply libraries/AP_Vehicle/AP_Vehicle.h:449]: you're suggested change has no actual change

[the remark this answers is not in the document]
Comment (tridge) [reply libraries/AP_Vehicle/AP_Vehicle.h:449]: the 30s wrap really doesn't matter, its not a precise number that matters in some way

**Comment by tridge** [comment]
> I'm a little surprised its possible to get a 2M build onto a 1M board and have it work well enough to print a warning.

yes, you can easily (I did it for testing). The 1M F427 boards have 2M flash, but have a (very small) chance of crashing when accessing USB while running code in 2nd 1M of flash

**Comment by tridge** [comment]
> I'd really just dump people into the config error loop in either of these cases...

that is not going to be a good user experience. I expect we have quite a few people using and flying like this. The crash with USB and 2M on a 1M board is rare, and it won't happen when flying as USB is not active

**Comment by peterbarker** [comment]
> > I'd really just dump people into the config error loop in either of these cases...
> 
> that is not going to be a good user experience. I expect we have quite a few people using and flying like this. The crash with USB and 2M on a 1M board is rare, and it won't happen when flying as USB is not active

Our thinking on the call was that the user had already just updated the firmware to a more recent version using e.g. MissionPlanner.  After update it should be relatively straight forward for them to re-flash it right there and then with a more-appropriate firmware.

**Review by peterbarker — APPROVED** [review]

**Pull request merged**